File: frontend/lib/finance.ts
import { api } from "@/lib/http";

export type OutstandingItem = {
  party: string;
  reference: string;
  due_date: string | null;
  amount: number;
};

export type FinanceSummary = {
  receivables: number;
  payables: number;
  net_position: number;
  overdue_receivables: number;
  outstanding_receivables: OutstandingItem[];
  outstanding_payables: OutstandingItem[];
};

export const getFinanceSummary = (token: string) =>
  api.get<FinanceSummary>(token, "/finance/summary");

// ---- Aged AR / AP ledger ----
export type AgingBuckets = {
  current: number;
  d30: number;
  d60: number;
  d90: number;
  older: number;
  total: number;
};
export type LedgerRow = {
  party: string;
  reference: string;
  posting_date: string | null;
  due_date: string | null;
  grand_total: number;
  outstanding: number;
  age_days: number;
  bucket: string;
};
export type LedgerResult = { rows: LedgerRow[]; totals: AgingBuckets };

export const getReceivable = (token: string) =>
  api.get<LedgerResult>(token, "/finance/receivable");
export const getPayable = (token: string) =>
  api.get<LedgerResult>(token, "/finance/payable");

// ---- Cash / bank book ----
export type BookEntry = {
  date: string | null;
  voucher_type: string | null;
  voucher_no: string | null;
  party: string | null;
  against: string | null;
  debit: number;
  credit: number;
  balance: number;
  remarks: string | null;
};
export type BookResult = {
  accounts: string[];
  opening: number;
  closing: number;
  total_debit: number;
  total_credit: number;
  entries: BookEntry[];
};

export const getCashBook = (token: string) =>
  api.get<BookResult>(token, "/finance/cash-book");
export const getBankBook = (token: string) =>
  api.get<BookResult>(token, "/finance/bank-book");

// ---- Financial statements ----
export type StatementLine = {
  account: string;
  indent: number;
  amount: number;
  is_total: boolean;
};
export type StatementResult = { title: string; rows: StatementLine[] };

export const getIncomeStatement = (
  token: string,
  params: { company: string; fiscal_year?: string }
) => api.get<StatementResult>(token, "/finance/reports/income-statement", params);

export const getBalanceSheet = (
  token: string,
  params: { company: string; fiscal_year?: string }
) => api.get<StatementResult>(token, "/finance/reports/balance-sheet", params);

// ---- Trial balance ----
export type TrialBalanceRow = { account: string; debit: number; credit: number };
export type TrialBalanceResult = {
  title: string;
  rows: TrialBalanceRow[];
  total_debit: number;
  total_credit: number;
};

export const getTrialBalance = (
  token: string,
  params: { company: string; fiscal_year?: string }
) => api.get<TrialBalanceResult>(token, "/finance/trial-balance", params);

// ---- Cash flow statement (direct method) ----
export type CashFlowLine = { label: string; amount: number };
export type CashFlowResult = {
  title: string;
  opening: number;
  closing: number;
  total_in: number;
  total_out: number;
  net_change: number;
  inflows: CashFlowLine[];
  outflows: CashFlowLine[];
};

export const getCashFlow = (
  token: string,
  params: { company?: string; fiscal_year?: string }
) => api.get<CashFlowResult>(token, "/finance/cash-flow", params);

File: backend/schemas/finance.py
"""Finance schemas (Pydantic V2): receivables/payables summary + report passthrough."""

from pydantic import BaseModel


class OutstandingItem(BaseModel):
    party: str
    reference: str
    due_date: str | None = None
    amount: float


class FinanceSummary(BaseModel):
    receivables: float
    payables: float
    net_position: float
    overdue_receivables: float
    outstanding_receivables: list[OutstandingItem]
    outstanding_payables: list[OutstandingItem]


# ---- Aged receivable / payable ledger ----
class AgingBuckets(BaseModel):
    current: float = 0    # not yet due
    d30: float = 0        # 1-30 days overdue
    d60: float = 0        # 31-60
    d90: float = 0        # 61-90
    # 91-120 was previously folded into `older`, which put a debt four months
    # late beside one a year late and gave collections nothing to sort on. The
    # boundary is 120 to match ERPNext's own ageing, so the two reports can be
    # read side by side.
    d120: float = 0       # 91-120
    older: float = 0      # over 120
    total: float = 0


class LedgerRow(BaseModel):
    party: str
    reference: str
    posting_date: str | None = None
    due_date: str | None = None
    grand_total: float
    outstanding: float
    age_days: int
    bucket: str  # "Current" | "1-30" | "31-60" | "61-90" | "91-120" | "120+"


class LedgerResult(BaseModel):
    rows: list[LedgerRow]
    totals: AgingBuckets


# ---- Cash / bank book (from the general ledger) ----
class BookEntry(BaseModel):
    date: str | None = None
    voucher_type: str | None = None
    voucher_no: str | None = None
    party: str | None = None
    against: str | None = None
    debit: float = 0
    credit: float = 0
    balance: float = 0
    remarks: str | None = None


class BookResult(BaseModel):
    accounts: list[str]
    opening: float
    closing: float
    total_debit: float
    total_credit: float
    entries: list[BookEntry]


# ---- Financial statements (normalised report rows) ----
class StatementLine(BaseModel):
    account: str
    indent: int = 0
    amount: float = 0
    is_total: bool = False


class StatementResult(BaseModel):
    title: str
    rows: list[StatementLine]


# ---- Trial balance (closing debit/credit per leaf account) ----
class TrialBalanceRow(BaseModel):
    account: str
    debit: float = 0
    credit: float = 0


class TrialBalanceResult(BaseModel):
    title: str = "Trial Balance"
    rows: list[TrialBalanceRow]
    total_debit: float = 0
    total_credit: float = 0


# ---- Cash flow statement (direct method, from the cash & bank ledger) ----
class CashFlowLine(BaseModel):
    label: str
    amount: float = 0


class CashFlowResult(BaseModel):
    title: str = "Cash Flow Statement"
    opening: float = 0
    closing: float = 0
    total_in: float = 0
    total_out: float = 0
    net_change: float = 0
    inflows: list[CashFlowLine]
    outflows: list[CashFlowLine]
